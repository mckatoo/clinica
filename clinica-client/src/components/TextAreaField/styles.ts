import styled, { css, DefaultTheme } from 'styled-components'

import { TextAreaFieldProps } from '.'

type WrapperProps = Pick<TextAreaFieldProps, 'disabled'> & { error?: boolean }

export const TextAreaWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`

export const TextArea = styled.textarea`
  ${({ theme }) => css`
    overflow: hidden;
    width: 100%;
    height: auto;
    resize: none;
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 ${theme.spacings.small}
        ${theme.colors.lightGray} inset;
      filter: none;
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.red};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${TextAreaWrapper} {
      border-color: ${theme.colors.red};
    }

    ${Label} {
      color: ${theme.colors.red};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${TextArea} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`
