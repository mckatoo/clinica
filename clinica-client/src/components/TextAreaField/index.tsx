import { TextareaHTMLAttributes, useState } from 'react'

import * as S from './styles'

export type TextAreaFieldProps = {
  onTextAreaChange?: (value: string) => void
  label?: string
  initialValue?: string
  disabled?: boolean
  error?: string
} & TextareaHTMLAttributes<HTMLTextAreaElement>

const TextAreaField = ({
  label,
  name,
  initialValue = '',
  error,
  disabled = false,
  onTextAreaChange,
  ...props
}: TextAreaFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const element = e.currentTarget
    const newValue = element.value
    element.style.height = '5px'
    element.style.width = '25px'
    element.style.height = `${element.scrollHeight}px`
    element.style.width = `${element.scrollHeight}px`
    setValue(newValue)

    !!onTextAreaChange && onTextAreaChange(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name!}>{label}</S.Label>}
      <S.TextAreaWrapper>
        <S.TextArea
          onChange={onChange}
          role='textbox'
          value={value}
          disabled={disabled}
          contentEditable={!disabled}
          name={name!}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.TextAreaWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default TextAreaField
