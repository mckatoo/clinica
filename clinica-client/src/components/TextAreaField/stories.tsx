import { Meta, Story } from '@storybook/react/types-6-0'

import TextAreaField, { TextAreaFieldProps } from './'

export default {
  title: 'Form/TextAreaField',
  component: TextAreaField,
  args: {
    label: 'Address',
    name: 'address',
    initialValue: '',
    placeholder: 'Endere√ßo do fulano \nCidade\nEstado\nCEP',
    disabled: false
  },
  argTypes: {
    onInput: { action: 'changed' }
  }
} as Meta

export const Default: Story<TextAreaFieldProps> = args => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextAreaField {...args} />
  </div>
)

export const withError: Story<TextAreaFieldProps> = args => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextAreaField {...args} />
  </div>
)

withError.args = {
  error: 'Ops...something is wrong'
}
