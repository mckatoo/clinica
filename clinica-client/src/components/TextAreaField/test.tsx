import { render, screen, waitFor } from 'utils/test-utils'

import userEvent from '@testing-library/user-event'

import TextAreaField from './'

describe('<TextAreaField />', () => {
  it('Renders with Label', () => {
    render(<TextAreaField label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<TextAreaField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<TextAreaField placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Changes its value when typing', async () => {
    const onTextAreaChange = jest.fn()
    render(
      <TextAreaField
        onTextAreaChange={onTextAreaChange}
        label="TextAreaField"
        name="TextAreaField"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onTextAreaChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onTextAreaChange).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onTextAreaChange = jest.fn()
    render(
      <TextAreaField
        onTextAreaChange={onTextAreaChange}
        label="TextAreaField"
        name="TextAreaField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onTextAreaChange).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <TextAreaField label="TextAreaField" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<TextAreaField label="TextAreaField" name="TextAreaField" />)

    const input = screen.getByLabelText('TextAreaField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(
      <TextAreaField label="TextAreaField" name="TextAreaField" disabled />
    )

    const input = screen.getByLabelText('TextAreaField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
