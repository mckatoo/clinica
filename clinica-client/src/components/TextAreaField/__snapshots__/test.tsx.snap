// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextAreaField /> Renders with error 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #ED3237;
}

.c5 {
  overflow: hidden;
  width: 100%;
  height: auto;
  resize: none;
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
}

.c5:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inset;
  -webkit-filter: none;
  filter: none;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  color: #EF4549;
  font-size: 1.2rem;
}

.c0 .c3 {
  border-color: #EF4549;
}

.c0 .c1 {
  color: #EF4549;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
  >
    TextAreaField
  </label>
  <div
    class="c3 c4"
  >
    <textarea
      class="c5"
      role="textbox"
    />
  </div>
  <p
    class="c6"
  >
    Error message
  </p>
</div>
`;
